kvs: improve append performance

chu11 opened the pull request:
As discussed in issue #2505, this fixes performance issue "part 1" by trying to reduce calls to `treeobj_validate()` (which calls `blobref_validate()`) on appends.  The constant interating through
append blobref arrays eats up tons of CPU time.

I was originally going to create a function like `treeobj_append_blobref_to_entry (dir, key, blobref)`, but as I implemented it, the function became stupid.  Basically just a cut & paste of:

```
treeobj_get_entry()
treeobj_deep_copy()
treeobj_append_blobref()
treeobj_insert_entry()
```

which was basically the the exact same code for appending in `kvstxn.c`, except I'd replace
`treeobj_insert_entry()` with an implementation without calling `treeobj_validate()`.

So I instead created a new function which is exactly the same as `treeobj_insert_entry()` without the `treeobj_validate()` call.  I call this brand spanking new function `treeobj_insert_entry2()`
... b/c I am not creative :-)  (I just couldn't come up with something I liked.  `treeobj_insert_entry_quick()`, `no_validate()`?)

Another option is to simply write a function internal to the KVS module that does the work of `treeobj_insert_entry2()` and just not make it public.  Tree objects are public knowledge in RFCs afterall.

Running `perf record src/cmd/flux start --size=64 flux mini run -n64 t/shell/lptest 80 500`

before

```
  18.69%        265648  flux-broker-0    kvs.so                     [.] blobref_validate
  12.26%        182796  flux-broker-0    libc-2.17.so               [.] _int_malloc
   8.82%        125456  flux-broker-0    libc-2.17.so               [.] malloc_consolidate
   6.00%         91512  flux-broker-0    libc-2.17.so               [.] _int_free
   5.10%        222610  lt-flux-shell    libflux-core.so.2.0.0      [.] cbuf_is_valid
   4.69%         67577  flux-broker-0    libc-2.17.so               [.] malloc
   2.91%         41611  flux-broker-0    libc-2.17.so               [.] __strcmp_sse42
   2.84%        124329  lt-flux-shell    libpthread-2.17.so         [.] pthread_mutex_trylock
   2.31%         32737  flux-broker-0    libjansson.so.4.10.0       [.] json_string_value
   2.29%         32631  flux-broker-0    libc-2.17.so               [.] __strlen_sse2_pminub
   1.86%         27390  flux-broker-0    libc-2.17.so               [.] free
```

after

```
  18.93%        175242  flux-broker-0    libc-2.17.so                  [.] _int_malloc
  14.00%        121656  flux-broker-0    libc-2.17.so                  [.] malloc_consolidate
   9.43%         92052  flux-broker-0    libc-2.17.so                  [.] _int_free
   7.68%         67754  flux-broker-0    libc-2.17.so                  [.] malloc
   4.54%        122239  lt-flux-shell    libflux-core.so.2.0.0         [.] cbuf_is_valid
   2.85%         26445  flux-broker-0    libc-2.17.so                  [.] free
   2.58%         69517  lt-flux-shell    libpthread-2.17.so            [.] pthread_mutex_trylock
   2.04%         17650  flux-broker-0    libjansson.so.4.10.0          [.] json_delete
   1.73%         15027  flux-broker-0    libjansson.so.4.10.0          [.] json_deep_copy
   1.64%         44402  lt-flux-shell    libflux-core.so.2.0.0         [.] epoll_poll
```

before and after run-times of just `src/cmd/flux start --size=64 flux mini run -n64 t/shell/lptest 80 500` using `time`.

before

`435.153u 339.150s 4:31.15 285.5%        0+0k 0+0io 0pf+0w`

after

`268.189u 290.647s 2:45.43 337.7%        0+0k 0+0io 0pf+0w`


not the the most accurate benchmark, but the `blobref_validate()` was clearly having an impact.

garlick commented:
Nice!  Instead of `*_2()` how about `*_novalidate()`?   Maybe not the prettiest name but it does hint at the difference compared to the other one.

chu11 commented:
> Nice! Instead of *_2() how about *_novalidate()? Maybe not the prettiest name but it does hint at the difference compared to the other one.

Since there's atleast one vote for `novalidate()`, we'll go with that one :-)

chu11 commented:
re-pushed, renaming `treeobj_insert_entry2()` to `treeobj_insert_entry_novalidate()`.  Also split the 1 commit into 2 and added a few unit tests to `libkvs/test/treeobj.c`.  Removed WIP, I think this is good to go now.
